### PR TITLE
Support SageMaker BYO finetuning

### PR DESCRIPTION
**Pull request by youran-qi**
This PR adds a new method called `export_finetune` to `Client` to support the SageMaker BYO finetuning.

This PR also adds a Jupyter Notebook to show how to export the customer's own finetuned merged weights to TensorRT-LLM engine, and deploy the endpoint from the exported TensorRT-LLM engine.

This PR has been tested from end to end: every cell in the Jupyter Notebook has been run and tested, all of which work well.

### CODE REVIEW

File: notebooks/sagemaker/Deploy your own finetuned command-r-0824.ipynb [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+    "## Pre-requisites:\n",
+    "\n",
+    "1. **Note: This notebook contains elements which render correctly in Jupyter interface. Open this notebook from an Amazon SageMaker Notebook Instance or Amazon SageMaker Studio.**\n",
+    "1. Ensure that IAM role used has **AmazonSageMakerFullAccess**\n",
+    "1. To deploy this ML model successfully, ensure that:\n",
+    "    1. Either your IAM role has these three permissions and you have authority to make AWS Marketplace subscriptions in the AWS account used: \n",
+    "        1. **aws-marketplace:ViewSubscriptions**\n",
+    "        1. **aws-marketplace:Unsubscribe**\n",
+    "        1. **aws-marketplace:Subscribe**  \n",
+    "    2. or your AWS account has a subscription to the packages for <span style=\"background-color: yellow\">Cohere Command R Bring Your Own Finetuning</span>. If so, skip step: [Subscribe to the bring your own finetuning algorithm](#1.-Subscribe-to-the-bring-your-own-finetuning-algorithm)\n",
```
Comment (youran-qi) [inline notebooks/sagemaker/Deploy your own finetuned command-r-0824.ipynb:25]: We need to add the URL of the Cohere Command R Bring Your Own Finetuning listing at here. I highlighted this part for now as a reminder.

File: notebooks/sagemaker/Deploy your own finetuned command-r-0824.ipynb [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "!pip install --upgrade cohere-aws"
```
Comment (youran-qi) [inline notebooks/sagemaker/Deploy your own finetuned command-r-0824.ipynb:83]: No longer need to install the `cohere-aws` after we migrate it to `cohere-python-sdk`

File: cohere_aws/client.py [hunk 3]
```diff
@@ -747,6 +747,72 @@ def create_finetune(
...
+        s3_merged_weights_dir: str,
+        s3_tensorrtllm_engine_dir: str,
+        arn: str,
+        instance_type: str = "ml.p4de.24xlarge",
+        role: Optional[str] = None,
+    ) -> None:
+        """Export the merged weights to the TensorRT-LLM inference engine.
+
+        Args:
+        name (str): The name to give to the export task.
```
Comment (CoderHam) [inline cohere_aws/client.py]:
```suggestion
        name (str): The name to while writing the exported model to the output directory.
```

File: cohere_aws/client.py [hunk 4]
```diff
@@ -747,6 +747,72 @@ def create_finetune(
         bucket, old_short_key = parse_s3_url(s3_models_dir + job_name)
         s3_resource.Bucket(bucket).objects.filter(Prefix=old_short_key).delete()
 
+    def export_finetune(
+        self,
+        name: str,
+        s3_merged_weights_dir: str,
+        s3_tensorrtllm_engine_dir: str,
```
Comment (CoderHam) [inline cohere_aws/client.py]:
```suggestion
        s3_output_dir: str,
```
Lets make it generic i.e. s3_output_dir

File: cohere_aws/client.py [hunk 5]
```diff
@@ -747,6 +747,72 @@ def create_finetune(
         bucket, old_short_key = parse_s3_url(s3_models_dir + job_name)
         s3_resource.Bucket(bucket).objects.filter(Prefix=old_short_key).delete()
 
+    def export_finetune(
+        self,
+        name: str,
+        s3_merged_weights_dir: str,
```
Comment (CoderHam) [inline cohere_aws/client.py]:
```suggestion
        s3_checkpoint_dir: str,
```
For folks doing non-PEFT finetunes merged weights can be confusing

File: cohere_aws/client.py [hunk 6]
```diff
@@ -747,6 +747,72 @@ def create_finetune(
...
+        instance_type (str, optional): The EC2 instance type to use for export. Defaults to "ml.p4de.24xlarge".
+        role (str, optional): The IAM role to use for export.
+            If not provided, sagemaker.get_execution_role() will be used to get the role.
+            This should work when one uses the client inside SageMaker. If this errors out,
+            the default role "ServiceRoleSagemaker" will be used, which generally works outside SageMaker.
+        """
+        if name == "model":
+            raise ValueError("name cannot be 'model'")
+
+        s3_tensorrtllm_engine_dir = s3_tensorrtllm_engine_dir.rstrip("/") + "/"
```
Comment (CoderHam) [inline cohere_aws/client.py]:
```suggestion
        s3_output_dir = s3_output_dir.rstrip("/") + "/"
```

File: cohere_aws/client.py [hunk 7]
```diff
@@ -747,6 +747,72 @@ def create_finetune(
...
+        s3_merged_weights_dir (str): An S3 path pointing to the directory of the merged weights.
+        s3_tensorrtllm_engine_dir (str): An S3 path pointing to the directory where the TensorRT-LLM engine will be saved.
+        arn (str): The product ARN of the bring your own finetuning algorithm.
+        instance_type (str, optional): The EC2 instance type to use for export. Defaults to "ml.p4de.24xlarge".
+        role (str, optional): The IAM role to use for export.
+            If not provided, sagemaker.get_execution_role() will be used to get the role.
+            This should work when one uses the client inside SageMaker. If this errors out,
+            the default role "ServiceRoleSagemaker" will be used, which generally works outside SageMaker.
+        """
+        if name == "model":
```
Comment (CoderHam) [inline cohere_aws/client.py:772]: Why prevent the name from being "model"?

File: cohere_aws/client.py [hunk 8]
```diff
@@ -747,6 +747,72 @@ def create_finetune(
...
+        s3_tensorrtllm_engine_dir = s3_tensorrtllm_engine_dir.rstrip("/") + "/"
+
+        if role is None:
+            try:
+                role = sage.get_execution_role()
+            except ValueError:
+                print("Using default role: 'ServiceRoleSagemaker'.")
+                role = "ServiceRoleSagemaker"
+
+        training_parameters = {"name": name}
```
Comment (CoderHam) [inline cohere_aws/client.py]:
```suggestion
        export_parameters = {"name": name}
```

File: cohere_aws/client.py [hunk 9]
```diff
@@ -747,6 +747,72 @@ def create_finetune(
...
+        inputs = {"checkpoint": s3_merged_weights_dir}
+
+        estimator.fit(inputs=inputs)
+
+        job_name = estimator.latest_training_job.name
+        current_filepath = f"{s3_tensorrtllm_engine_dir}{job_name}/output/model.tar.gz"
+
+        s3_resource = boto3.resource("s3")
+
+        # Copy the exported TensorRT-LLM engine to the root of s3_tensorrtllm_engine_dir
```
Comment (CoderHam) [inline cohere_aws/client.py]: Why is this needed?

**Review by CoderHam — CHANGES_REQUESTED** [review]

File: cohere_aws/client.py (hunk 3 above)
Comment (youran-qi) [reply cohere_aws/client.py]: Done. See [here](https://github.com/cohere-ai/cohere-aws/pull/168/commits/c946a2c798baf4fb42a3ecfc29f197ba62ee9586)

File: cohere_aws/client.py (hunk 4 above)
Comment (youran-qi) [reply cohere_aws/client.py]: Done. See [here](https://github.com/cohere-ai/cohere-aws/pull/168/commits/c946a2c798baf4fb42a3ecfc29f197ba62ee9586)

File: cohere_aws/client.py (hunk 5 above)
Comment (youran-qi) [reply cohere_aws/client.py]: Done. See [here](https://github.com/cohere-ai/cohere-aws/pull/168/commits/c946a2c798baf4fb42a3ecfc29f197ba62ee9586)

File: cohere_aws/client.py (hunk 6 above)
Comment (youran-qi) [reply cohere_aws/client.py]: Done. See [here](https://github.com/cohere-ai/cohere-aws/pull/168/commits/c946a2c798baf4fb42a3ecfc29f197ba62ee9586)

File: cohere_aws/client.py (hunk 8 above)
Comment (youran-qi) [reply cohere_aws/client.py]: Done. See [here](https://github.com/cohere-ai/cohere-aws/pull/168/commits/c946a2c798baf4fb42a3ecfc29f197ba62ee9586)

File: cohere_aws/client.py (hunk 9 above)
Comment (youran-qi) [reply cohere_aws/client.py]: This will make the result (TRT-LLM engine) more clean. 

Without this, customers will find the TRT-LLM engine in a path like `{s3_output_dir}/{name}-2024-09-20-22-04-35-757/output/model.tar.gz`. With this, they will find it in a path like `{s3_output_dir}/{name}.tar.gz`, which is more clean.

File: cohere_aws/client.py (hunk 7 above)
Comment (youran-qi) [reply cohere_aws/client.py:772]: I added this because I see https://github.com/cohere-ai/cohere-aws/blame/main/cohere_aws/client.py#L701, which I guess it is because `model.tar.gz` is some reserved name for our internal use, but I am not sure. 

Removing it probably won't cause issues, but putting it here is not harmful, so I decided to put it here. I can remove this if you want.

Comment (CoderHam) [reply cohere_aws/client.py:772]: Ah, I think its fine to keep for now 😄

**Review by CoderHam — APPROVED** [review]
